File: packages/core/src/guard.ts
import { createHash, randomUUID } from 'node:crypto'
import type { GuardOptions, ResolvedOptions } from './types.js'
import { UsageTracker } from './budget-tracker.js'
import { ResourceLimitTracker } from './resource-limit-tracker.js'
import { LoopDetector } from './loop-detector.js'
import { SideEffectRegistry } from './side-effect-registry.js'
import { TraceRecorder } from './trace-recorder.js'
import type { FuzeService } from './services/index.js'
import { extractUsageFromResult } from './usage-extractor.js'
import { LoopDetected, GuardTimeout, FuzeError, ResourceLimitExceeded } from './errors.js'

/**
 * Internal context shared across all guarded functions in a run.
 */
export interface GuardContext {
  runId: string
  usageTracker: UsageTracker
  resourceLimitTracker: ResourceLimitTracker
  loopDetector: LoopDetector
  sideEffectRegistry: SideEffectRegistry
  traceRecorder: TraceRecorder
  stepNumber: number
  /** Service for telemetry + remote config - NoopService when no daemon/cloud configured. */
  service: FuzeService
}

/**
 * Creates the guard wrapper function bound to a specific run context.
 * @param resolvedOpts - Fully resolved options for this run.
 * @param context - Shared run context (usage, loop, trace, side-effects).
 * @returns A function that wraps any sync/async function with protection.
 */
export function createGuardWrapper(resolvedOpts: ResolvedOptions, context: GuardContext) {
  /**
   * Wraps a sync or async function with Fuze protection:
   * loop detection, timeout, side-effect tracking, and trace recording.
   *
   * @param fn - The function to wrap.
   * @param options - Per-function guard options that override run-level config.
   * @returns A wrapped function with the same signature.
   */
  return function guard<T extends (...args: unknown[]) => unknown>(
    fn: T,
    options?: GuardOptions,
  ): T {
    const funcName = fn.name || 'anonymous'
    const opts = mergeStepOptions(resolvedOpts, options)

    // Register compensation if provided
    if (opts.compensate) {
      context.sideEffectRegistry.registerCompensation(funcName, opts.compensate)
    }

    const wrapped = async function (this: unknown, ...args: unknown[]): Promise<unknown> {
      const stepId = randomUUID()
      const argsHash = hashArgs(args)
      const startedAt = new Date().toISOString()
      const startMs = Date.now()
      const stepNumber = ++context.stepNumber

      try {
        await context.resourceLimitTracker.checkAndReserveStep(funcName)
      } catch (limitError) {
        if (limitError instanceof ResourceLimitExceeded) {
          context.traceRecorder.recordGuardEvent({
            eventId: randomUUID(),
            runId: context.runId,
            stepId,
            timestamp: new Date().toISOString(),
            type: 'kill',
            severity: 'critical',
            details: { ...limitError.details },
          })
          fireAndForget(context.service.sendGuardEvent(context.runId, {
            stepId,
            eventType: 'kill',
            severity: 'critical',
            details: { ...limitError.details, cause: 'resource_limit_exceeded' },
          }))
          await context.traceRecorder.flush()
        }
        throw limitError
      }

      // 1. Check loop detector - Layer 1: iteration cap
      const loopSignal = context.loopDetector.onStep()
      if (loopSignal) {
        context.traceRecorder.recordGuardEvent({
          eventId: randomUUID(),
          runId: context.runId,
          stepId,
          timestamp: new Date().toISOString(),
          type: 'loop_detected',
          severity: 'critical',
          details: loopSignal.details,
        })
        fireAndForget(context.service.sendGuardEvent(context.runId, {
          stepId, eventType: 'loop_detected', severity: 'critical', details: loopSignal.details,
        }))
        await context.traceRecorder.flush()

        if (opts.onLoop === 'kill') throw new LoopDetected(loopSignal, funcName)
        if (opts.onLoop === 'skip') return undefined
        // 'warn' - continue execution
      }

      // Layer 2: repeated tool call detection
      const toolSignature = `${funcName}:${argsHash}`
      const toolSignal = context.loopDetector.onToolCall(toolSignature)
      if (toolSignal) {
        context.traceRecorder.recordGuardEvent({
          eventId: randomUUID(),
          runId: context.runId,
          stepId,
          timestamp: new Date().toISOString(),
          type: 'loop_detected',
          severity: 'action',
          details: toolSignal.details,
        })
        fireAndForget(context.service.sendGuardEvent(context.runId, {
          stepId, eventType: 'loop_detected', severity: 'action', details: toolSignal.details,
        }))

        if (opts.onLoop === 'kill') {
          await context.traceRecorder.flush()
          throw new LoopDetected(toolSignal, funcName)
        }
        if (opts.onLoop === 'skip') return undefined
      }

      // Apply remote config overrides (synchronous cache read - zero latency)
      let callOpts = { ...opts }
      const remoteConfig = context.service.getToolConfig(funcName)
      if (remoteConfig) {
        if (!remoteConfig.enabled) {
          throw new FuzeError(`Tool '${funcName}' is disabled via remote configuration`)
        }
        callOpts = {
          ...callOpts,
          maxRetries: remoteConfig.maxRetries,
          timeout: remoteConfig.timeout,
        }
      }

      // Check service (kill switch). Falls back to proceed if unavailable.
      const decision = await context.service.sendStepStart(context.runId, {
        stepId,
        stepNumber,
        toolName: funcName,
        argsHash,
        sideEffect: opts.sideEffect,
      })
      if (decision === 'kill') {
        throw new FuzeError('Transport kill: kill switch activated')
      }

      // 2. Execute with timeout + retry policy
      let result: unknown
      let error: string | undefined
      let timer: ReturnType<typeof setTimeout> | undefined
      let recordedTokensIn = 0
      let recordedTokensOut = 0

      const clearTimer = (): void => {
        if (timer !== undefined) {
          clearTimeout(timer)
          timer = undefined
        }
      }

      try {
        let attempt = 0
        let retriesRemaining = Math.max(0, callOpts.maxRetries)

        while (true) {
          attempt += 1
          try {
            if (callOpts.timeout < Infinity) {
              let timedOut = false
              const fnPromise = Promise.resolve(fn.apply(this, args))
                .catch((err: unknown) => {
                  if (timedOut) return undefined
                  throw err
                })
                .finally(() => {
                  clearTimer()
                })

              const timeoutPromise = new Promise<never>((_, reject) => {
                timer = setTimeout(() => {
                  timedOut = true
                  reject(new GuardTimeout(funcName, callOpts.timeout))
                }, callOpts.timeout)
              })

              result = await Promise.race([
                fnPromise,
                timeoutPromise,
              ])
            } else {
              result = await Promise.resolve(fn.apply(this, args))
            }

            break
          } catch (attemptError) {
            if (isNonRetryableError(attemptError) || retriesRemaining <= 0) {
              throw attemptError
            }

            retriesRemaining -= 1
            const backoffMs = Math.min(100 * (2 ** (attempt - 1)), 5000)
            const retryDetails = {
              attempt,
              nextAttempt: attempt + 1,
              retriesRemaining,
              maxRetries: callOpts.maxRetries,
              delayMs: backoffMs,
              error: attemptError instanceof Error ? attemptError.message : String(attemptError),
            }

            context.traceRecorder.recordGuardEvent({
              eventId: randomUUID(),
              runId: context.runId,
              stepId,
              timestamp: new Date().toISOString(),
              type: 'retry',
              severity: 'warning',
              details: retryDetails,
            })
            fireAndForget(context.service.sendGuardEvent(context.runId, {
              stepId,
              eventType: 'retry',
              severity: 'warning',
              details: retryDetails,
            }))

            await wait(backoffMs)
          } finally {
            clearTimer()
          }
        }
      } catch (err) {
        error = err instanceof Error ? err.message : String(err)
        throw err
      } finally {
        clearTimer()

        // 3. Extract token usage from result (auto-detection or custom extractor)
        const endedAt = new Date().toISOString()
        const latencyMs = Date.now() - startMs

        const extracted = result !== undefined
          ? (callOpts.usageExtractor ? callOpts.usageExtractor(result) : extractUsageFromResult(result))
          : null

        if (extracted) {
          recordedTokensIn = extracted.tokensIn
          recordedTokensOut = extracted.tokensOut
        }

        context.traceRecorder.recordStep({
          stepId,
          runId: context.runId,
          stepNumber,
          startedAt,
          endedAt,
          toolName: funcName,
          argsHash,
          hasSideEffect: opts.sideEffect,
          tokensIn: recordedTokensIn,
          tokensOut: recordedTokensOut,
          latencyMs,
          error,
        })

        context.usageTracker.recordUsage(recordedTokensIn, recordedTokensOut)
        context.resourceLimitTracker.recordUsage(recordedTokensIn, recordedTokensOut)

        // Notify transport of step completion (fire-and-forget)
        fireAndForget(context.service.sendStepEnd(context.runId, stepId, {
          toolName: funcName,
          stepNumber,
          argsHash,
          hasSideEffect: opts.sideEffect,
          tokensIn: recordedTokensIn,
          tokensOut: recordedTokensOut,
          latencyMs,
          error: error ?? null,
        }))
      }

      // 4. Check progress - Layer 3 (no-progress detection)
      const hasNewOutput = result !== undefined && result !== null
      const progressSignal = context.loopDetector.onProgress(hasNewOutput)
      if (progressSignal) {
        context.traceRecorder.recordGuardEvent({
          eventId: randomUUID(),
          runId: context.runId,
          stepId: stepId,
          timestamp: new Date().toISOString(),
          type: 'loop_detected',
          severity: 'warning',
          details: progressSignal.details,
        })
        fireAndForget(context.service.sendGuardEvent(context.runId, {
          stepId, eventType: 'loop_detected', severity: 'warning', details: progressSignal.details,
        }))

        if (opts.onLoop === 'kill') {
          await context.traceRecorder.flush()
          throw new LoopDetected(progressSignal, funcName)
        }
      }

      // 5. Record side-effect if applicable
      if (opts.sideEffect) {
        context.sideEffectRegistry.recordSideEffect(stepId, funcName, result)
      }

      return result
    }

    // Preserve function name for debugging
    Object.defineProperty(wrapped, 'name', { value: funcName, configurable: true })

    return wrapped as unknown as T
  }
}

// Match Python's _hash_args: serialize as `{"args": [...], "kwargs": {...}}`
// with deep-sorted keys and compact separators, so positional-only JS calls
// produce the same hash bytes as Python `*args` calls with no kwargs. JS has
// no kwargs, so the kwargs object is always empty.
function canonicalStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    if (typeof value === 'undefined') return 'null'
    return JSON.stringify(value)
  }
  if (Array.isArray(value)) {
    return '[' + value.map(canonicalStringify).join(',') + ']'
  }
  const obj = value as Record<string, unknown>
  const keys = Object.keys(obj).sort()
  return '{' + keys.map((k) => JSON.stringify(k) + ':' + canonicalStringify(obj[k])).join(',') + '}'
}

function hashArgs(args: unknown[]): string {
  const hash = createHash('sha256')
  let serialized = '[unserializable]'
  try {
    serialized = canonicalStringify({ args, kwargs: {} })
  } catch {
    serialized = String(args)
  }
  hash.update(serialized)
  return hash.digest('hex').slice(0, 16)
}

function fireAndForget(promise: Promise<unknown>): void {
  promise.catch(() => undefined)
}

function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

function isNonRetryableError(err: unknown): boolean {
  return err instanceof GuardTimeout || err instanceof LoopDetected
}

/**
 * Merges step-level options with resolved run-level options.
 */
export function mergeStepOptions(
  resolved: ResolvedOptions,
  step?: GuardOptions,
): ResolvedOptions {
  if (!step) return resolved

  return {
    ...resolved,
    maxRetries: step.maxRetries ?? resolved.maxRetries,
    timeout: step.timeout ?? resolved.timeout,
    maxIterations: step.maxIterations ?? resolved.maxIterations,
    onLoop: step.onLoop ?? resolved.onLoop,
    traceOutput: resolved.traceOutput,
    sideEffect: step.sideEffect ?? resolved.sideEffect,
    compensate: step.compensate ?? resolved.compensate,
    usageExtractor: step.usageExtractor ?? resolved.usageExtractor,
    loopDetection: { ...resolved.loopDetection, ...step.loopDetection },
  }
}

File: packages/python/src/fuze_ai/guard.py
from __future__ import annotations

import asyncio
import functools
import hashlib
import inspect
import json
import math
import time
import uuid
from concurrent.futures import ThreadPoolExecutor, TimeoutError as FuturesTimeoutError
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, Optional, TypeVar

from fuze_ai.config_loader import ConfigLoader
from fuze_ai.errors import FuzeError, GuardTimeout, LoopDetected, ResourceLimitExceeded
from fuze_ai.loop_detector import LoopDetector
from fuze_ai.pricing import extract_usage_from_result
from fuze_ai.resource_limit_tracker import ResourceLimitTracker
from fuze_ai.services.types import FuzeService
from fuze_ai.side_effect_registry import SideEffectRegistry
from fuze_ai.trace_recorder import TraceRecorder
from fuze_ai.types import GuardEventRecord, ResolvedOptions, StepRecord

F = TypeVar("F", bound=Callable[..., Any])


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _hash_args(args: tuple[Any, ...], kwargs: dict[str, Any]) -> str:
    def _typed_default(value: Any) -> str:
        return f"{type(value).__name__}:{value}"

    payload = json.dumps({"args": args, "kwargs": kwargs}, default=_typed_default, sort_keys=True)
    return hashlib.sha256(payload.encode()).hexdigest()[:16]


async def _swallow(coro: Any) -> None:
    try:
        await coro
    except Exception:
        return


def _fire_and_forget(coro: Any) -> None:
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        try:
            asyncio.run(_swallow(coro))
        except Exception:
            return
    else:
        loop.create_task(_swallow(coro))


def _run_service_call(coro: Any, default: Any) -> Any:
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        try:
            return asyncio.run(coro)
        except Exception:
            return default

    with ThreadPoolExecutor(max_workers=1) as executor:
        future = executor.submit(asyncio.run, coro)
        try:
            return future.result(timeout=0.2)
        except Exception:
            return default


def _run_with_timeout(fn: Callable[..., Any], args: tuple, kwargs: dict, timeout_ms: float) -> Any:
    if not math.isfinite(timeout_ms) or timeout_ms <= 0:
        return fn(*args, **kwargs)

    timeout_s = timeout_ms / 1000.0
    executor = ThreadPoolExecutor(max_workers=1)
    future = executor.submit(fn, *args, **kwargs)
    try:
        return future.result(timeout=timeout_s)
    except FuturesTimeoutError as exc:
        future.cancel()
        raise GuardTimeout(getattr(fn, "__name__", "unknown"), int(timeout_ms)) from exc
    finally:
        try:
            executor.shutdown(wait=False, cancel_futures=True)
        except TypeError:
            executor.shutdown(wait=False)


async def _run_async_with_timeout(
    fn: Callable[..., Any], args: tuple, kwargs: dict, timeout_ms: float
) -> Any:
    if not math.isfinite(timeout_ms) or timeout_ms <= 0:
        return await fn(*args, **kwargs)

    timeout_s = timeout_ms / 1000.0
    try:
        return await asyncio.wait_for(fn(*args, **kwargs), timeout=timeout_s)
    except asyncio.TimeoutError as exc:
        raise GuardTimeout(getattr(fn, "__name__", "unknown"), int(timeout_ms)) from exc


@dataclass
class GuardContext:
    run_id: str
    resource_limit_tracker: ResourceLimitTracker
    loop_detector: LoopDetector
    side_effect_registry: SideEffectRegistry
    trace_recorder: TraceRecorder
    service: FuzeService
    step_number: int = field(default=0)


def _record_limit_event(
    ctx: GuardContext,
    step_id: str,
    err: ResourceLimitExceeded,
) -> None:
    details = {
        "tool_name": err.tool_name,
        "limit": err.limit,
        "ceiling": err.ceiling,
        "observed": err.observed,
    }
    ctx.trace_recorder.record_guard_event(GuardEventRecord(
        event_id=str(uuid.uuid4()),
        run_id=ctx.run_id,
        step_id=step_id,
        timestamp=_now_iso(),
        type="kill",
        severity="critical",
        details=details,
    ))
    _fire_and_forget(ctx.service.send_guard_event(
        ctx.run_id,
        {
            "step_id": step_id,
            "event_type": "kill",
            "severity": "critical",
            "details": {**details, "cause": "resource_limit_exceeded"},
        },
    ))
    ctx.trace_recorder.flush()


def _make_wrapper(fn: Callable[..., Any], opts: ResolvedOptions, ctx: GuardContext) -> Callable[..., Any]:
    func_name = fn.__name__ or "anonymous"

    if opts.get("compensate"):
        ctx.side_effect_registry.register_compensation(func_name, opts["compensate"])

    def _handle_loop_signal(signal: Any, step_id: str) -> None:
        ctx.trace_recorder.record_guard_event(GuardEventRecord(
            event_id=str(uuid.uuid4()),
            run_id=ctx.run_id,
            step_id=step_id,
            timestamp=_now_iso(),
            type="loop_detected",
            severity="critical",
            details=signal["details"],
        ))
        _fire_and_forget(ctx.service.send_guard_event(
            ctx.run_id,
            {
                "step_id": step_id,
                "event_type": "loop_detected",
                "severity": "critical",
                "details": signal["details"],
            },
        ))
        if opts["on_loop"] == "kill":
            ctx.trace_recorder.flush()
            raise LoopDetected(signal, func_name)

    if inspect.iscoroutinefunction(fn):
        @functools.wraps(fn)
        async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
            step_id = str(uuid.uuid4())
            args_hash = _hash_args(args, kwargs)
            start_time = time.monotonic()
            started_at = _now_iso()
            error_msg: Optional[str] = None
            result: Any = None

            try:
                await ctx.resource_limit_tracker.check_and_reserve_step(func_name)
            except ResourceLimitExceeded as err:
                _record_limit_event(ctx, step_id, err)
                raise

            ctx.step_number += 1
            step_number = ctx.step_number

            loop_signal = ctx.loop_detector.on_step()
            if loop_signal:
                _handle_loop_signal(loop_signal, step_id)
                if opts["on_loop"] == "skip":
                    return None

            tool_signal = ctx.loop_detector.on_tool_call(f"{func_name}:{args_hash}")
            if tool_signal:
                _handle_loop_signal(tool_signal, step_id)
                if opts["on_loop"] == "skip":
                    return None

            decision = await ctx.service.send_step_start(
                ctx.run_id,
                {
                    "step_id": step_id,
                    "step_number": step_number,
                    "tool_name": func_name,
                    "args_hash": args_hash,
                    "side_effect": bool(opts.get("side_effect")),
                },
            )
            if decision == "kill":
                raise FuzeError("Transport kill: kill switch activated")

            try:
                result = await _run_async_with_timeout(fn, args, kwargs, opts["timeout"])
            except (LoopDetected, GuardTimeout):
                raise
            except Exception as exc:
                error_msg = str(exc)
                raise
            finally:
                ended_at = _now_iso()
                latency_ms = int((time.monotonic() - start_time) * 1000)
                extractor = opts.get("usage_extractor") or extract_usage_from_result
                extracted = extractor(result) if result is not None else None

                if extracted:
                    actual_tokens_in = int(extracted.get("tokens_in") or 0)
                    actual_tokens_out = int(extracted.get("tokens_out") or 0)
                else:
                    actual_tokens_in = 0
                    actual_tokens_out = 0

                ctx.trace_recorder.record_step(StepRecord(
                    step_id=step_id,
                    run_id=ctx.run_id,
                    step_number=step_number,
                    started_at=started_at,
                    ended_at=ended_at,
                    tool_name=func_name,
                    args_hash=args_hash,
                    has_side_effect=opts.get("side_effect", False),
                    tokens_in=actual_tokens_in,
                    tokens_out=actual_tokens_out,
                    latency_ms=latency_ms,
                    error=error_msg,
                ))
                ctx.resource_limit_tracker.record_usage(actual_tokens_in, actual_tokens_out)
                _fire_and_forget(ctx.service.send_step_end(
                    ctx.run_id,
                    step_id,
                    {
                        "tool_name": func_name,
                        "step_number": step_number,
                        "args_hash": args_hash,
                        "has_side_effect": bool(opts.get("side_effect")),
                        "tokens_in": actual_tokens_in,
                        "tokens_out": actual_tokens_out,
                        "latency_ms": latency_ms,
                        "error": error_msg,
                    },
                ))

            has_new_output = result is not None
            progress_signal = ctx.loop_detector.on_progress(has_new_output)
            if progress_signal:
                ctx.trace_recorder.record_guard_event(GuardEventRecord(
                    event_id=str(uuid.uuid4()),
                    run_id=ctx.run_id,
                    step_id=step_id,
                    timestamp=_now_iso(),
                    type="loop_detected",
                    severity="warning",
                    details=progress_signal["details"],
                ))
                _fire_and_forget(ctx.service.send_guard_event(
                    ctx.run_id,
                    {
                        "step_id": step_id,
                        "event_type": "loop_detected",
                        "severity": "warning",
                        "details": progress_signal["details"],
                    },
                ))
                if opts["on_loop"] == "kill":
                    ctx.trace_recorder.flush()
                    raise LoopDetected(progress_signal, func_name)

            if opts.get("side_effect"):
                ctx.side_effect_registry.record_side_effect(step_id, func_name, result)

            return result

        return async_wrapper

    @functools.wraps(fn)
    def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
        step_id = str(uuid.uuid4())
        args_hash = _hash_args(args, kwargs)
        start_time = time.monotonic()
        started_at = _now_iso()
        error_msg: Optional[str] = None
        result: Any = None

        try:
            ctx.resource_limit_tracker.check_and_reserve_step_sync(func_name)
        except ResourceLimitExceeded as err:
            _record_limit_event(ctx, step_id, err)
            raise

        ctx.step_number += 1
        step_number = ctx.step_number

        loop_signal = ctx.loop_detector.on_step()
        if loop_signal:
            _handle_loop_signal(loop_signal, step_id)
            if opts["on_loop"] == "skip":
                return None

        tool_signal = ctx.loop_detector.on_tool_call(f"{func_name}:{args_hash}")
        if tool_signal:
            _handle_loop_signal(tool_signal, step_id)
            if opts["on_loop"] == "skip":
                return None

        decision = _run_service_call(
            ctx.service.send_step_start(
                ctx.run_id,
                {
                    "step_id": step_id,
                    "step_number": step_number,
                    "tool_name": func_name,
                    "args_hash": args_hash,
                    "side_effect": bool(opts.get("side_effect")),
                },
            ),
            "proceed",
        )
        if decision == "kill":
            raise FuzeError("Transport kill: kill switch activated")

        try:
            result = _run_with_timeout(fn, args, kwargs, opts["timeout"])
        except (LoopDetected, GuardTimeout):
            raise
        except Exception as exc:
            error_msg = str(exc)
            raise
        finally:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - start_time) * 1000)
            extractor = opts.get("usage_extractor") or extract_usage_from_result
            extracted = extractor(result) if result is not None else None

            if extracted:
                actual_tokens_in = int(extracted.get("tokens_in") or 0)
                actual_tokens_out = int(extracted.get("tokens_out") or 0)
            else:
                actual_tokens_in = 0
                actual_tokens_out = 0

            ctx.trace_recorder.record_step(StepRecord(
                step_id=step_id,
                run_id=ctx.run_id,
                step_number=step_number,
                started_at=started_at,
                ended_at=ended_at,
                tool_name=func_name,
                args_hash=args_hash,
                has_side_effect=opts.get("side_effect", False),
                tokens_in=actual_tokens_in,
                tokens_out=actual_tokens_out,
                latency_ms=latency_ms,
                error=error_msg,
            ))
            ctx.resource_limit_tracker.record_usage(actual_tokens_in, actual_tokens_out)
            _fire_and_forget(ctx.service.send_step_end(
                ctx.run_id,
                step_id,
                {
                    "tool_name": func_name,
                    "step_number": step_number,
                    "args_hash": args_hash,
                    "has_side_effect": bool(opts.get("side_effect")),
                    "tokens_in": actual_tokens_in,
                    "tokens_out": actual_tokens_out,
                    "latency_ms": latency_ms,
                    "error": error_msg,
                },
            ))

        has_new_output = result is not None
        progress_signal = ctx.loop_detector.on_progress(has_new_output)
        if progress_signal:
            ctx.trace_recorder.record_guard_event(GuardEventRecord(
                event_id=str(uuid.uuid4()),
                run_id=ctx.run_id,
                step_id=step_id,
                timestamp=_now_iso(),
                type="loop_detected",
                severity="warning",
                details=progress_signal["details"],
            ))
            _fire_and_forget(ctx.service.send_guard_event(
                ctx.run_id,
                {
                    "step_id": step_id,
                    "event_type": "loop_detected",
                    "severity": "warning",
                    "details": progress_signal["details"],
                },
            ))
            if opts["on_loop"] == "kill":
                ctx.trace_recorder.flush()
                raise LoopDetected(progress_signal, func_name)

        if opts.get("side_effect"):
            ctx.side_effect_registry.record_side_effect(step_id, func_name, result)

        return result

    return sync_wrapper


def _build_context(opts: ResolvedOptions, service: FuzeService) -> GuardContext:
    return GuardContext(
        run_id=str(uuid.uuid4()),
        resource_limit_tracker=ResourceLimitTracker(opts.get("resource_limits", {})),
        loop_detector=LoopDetector(
            max_iterations=opts["max_iterations"],
            window_size=opts["loop_detection"]["window_size"],
            repeat_threshold=opts["loop_detection"]["repeat_threshold"],
            max_flat_steps=opts["loop_detection"]["max_flat_steps"],
        ),
        side_effect_registry=SideEffectRegistry(),
        trace_recorder=TraceRecorder(opts["trace_output"]),
        service=service,
    )
